File: addons/tr03112/src/jvmMain/java/org/openecard/binding/tctoken/TCToken.java
/****************************************************************************
 * Copyright (C) 2014-2016 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.binding.tctoken;

import org.openecard.binding.tctoken.ex.InvalidRedirectUrlException;
import generated.TCTokenType;
import java.net.URI;
import java.net.URISyntaxException;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import static org.openecard.binding.tctoken.ex.ErrorTranslations.*;
import org.openecard.common.util.UrlBuilder;


/**
 * Helper class adding further functionality to the underlying TCTokenType.
 *
 * @author Tobias Wich
 */
public class TCToken extends TCTokenType {

    private static final Logger LOG = LoggerFactory.getLogger(TCToken.class);

    private boolean invalidPSK = false;

    /**
     * Gets the CommunicationErrorAddress for use in error conditions.
     * If the CommunicationErrorAddress is available this one is used.
     *
     * @param minor The ResultMinor string.
     * @return The error URL.
     * @throws InvalidRedirectUrlException In case the address is not present or a valid URL.
     */
    public String getComErrorAddressWithParams(@Nonnull String minor) throws InvalidRedirectUrlException {
	try {
	    String errorUrl = getCommunicationErrorAddress();
	    URI url = checkUrl(errorUrl);
	    String result = UrlBuilder.fromUrl(url).queryParam("ResultMajor", "error")
		    .queryParamUrl("ResultMinor", TCTokenHacks.fixResultMinor(minor))
		    .build().toString();
	    return result;
	} catch (URISyntaxException ex) {
	    // should not happen, but here it is anyways
	    LOG.error("Construction of redirect URL failed.", ex);
	    throw new InvalidRedirectUrlException(NO_URL);
	}
    }

    /**
     * Indicates whether the original data of the TCToken contains a invalid PSK.
     * <br>
     * Note: This method is just for the case there was PSK data in the original data received. If the original TCToken
     * data does not contain the PathSecurity-Parameters or the PSK element than this method will return false.
     *
     * @return {@code TRUE} if the PSK contain in the TCToken is invalid else {@code FALSE}.
     */
    protected boolean isInvalidPSK() {
	return invalidPSK;
    }

    /**
     * Sets the indicator of an invalid PSK.
     * <br>
     * Note: This method is just for the case there is a PSK in the original TCToken but it is invalid for some reason.
     * If there is no PSK in the TCToken data this message should not be used.
     *
     * @param invalidPSK {@code TRUE} if the PSK should be marked as invalid else {@code FALSE}. The indicator is set to
     * {@code FALSE} per default.
     */
    protected void setInvalidPSK(boolean invalidPSK) {
	this.invalidPSK = invalidPSK;
    }

    private static URI checkUrl(@Nullable String urlStr) throws InvalidRedirectUrlException {
	if (urlStr != null && ! urlStr.isEmpty()) {
	    try {
		URI url = new URI(urlStr);
		return url;
	    } catch (URISyntaxException ex) {
		LOG.error("No valid CommunicationErrorAddress provided.");
		throw new InvalidRedirectUrlException(NO_URL);
	    }
	} else {
	    LOG.error("No CommunicationErrorAddress to perform a redirect provided.");
	    throw new InvalidRedirectUrlException(NO_REDIRECT_AVAILABLE);
	}
    }

}

File: addons/tr03112/src/jvmMain/kotlin/org/openecard/binding/tctoken/TCToken.kt
/****************************************************************************
 * Copyright (C) 2014-2016 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 */
package org.openecard.binding.tctoken

import generated.TCTokenType
import io.github.oshai.kotlinlogging.KotlinLogging
import org.openecard.binding.tctoken.ex.ErrorTranslations
import org.openecard.binding.tctoken.ex.InvalidRedirectUrlException
import org.openecard.common.util.UrlBuilder
import java.net.URI
import java.net.URISyntaxException

private val logger = KotlinLogging.logger { }

/**
 * Helper class adding further functionality to the underlying TCTokenType.
 *
 * @author Tobias Wich
 */
class TCToken : TCTokenType() {
	/**
	 * Sets the indicator of an invalid PSK.
	 * <br></br>
	 * Note: This method is just for the case there is a PSK in the original TCToken but it is invalid for some reason.
	 * If there is no PSK in the TCToken data this message should not be used.
	 *
	 * @param isInvalidPSK `TRUE` if the PSK should be marked as invalid else `FALSE`. The indicator is set to
	 * `FALSE` per default.
	 */
	var isInvalidPSK: Boolean = false

	/**
	 * Gets the CommunicationErrorAddress for use in error conditions.
	 * If the CommunicationErrorAddress is available this one is used.
	 *
	 * @param minor The ResultMinor string.
	 * @return The error URL.
	 * @throws InvalidRedirectUrlException In case the address is not present or a valid URL.
	 */
	@Throws(InvalidRedirectUrlException::class)
	fun getComErrorAddressWithParams(minor: String): String {
		try {
			val errorUrl = getCommunicationErrorAddress()
			val url: URI = checkUrl(errorUrl)
			val result: String =
				UrlBuilder
					.fromUrl(url)
					.queryParam("ResultMajor", "error")
					.queryParamUrl("ResultMinor", TCTokenHacks.fixResultMinor(minor))
					.build()
					.toString()
			return result
		} catch (ex: URISyntaxException) {
			// should not happen, but here it is anyways
			logger.error(ex) { "Construction of redirect URL failed." }
			throw InvalidRedirectUrlException(ErrorTranslations.NO_URL)
		}
	}

	companion object {
		private fun checkUrl(urlStr: String?): URI {
			if (urlStr != null && !urlStr.isEmpty()) {
				try {
					val url = URI(urlStr)
					return url
				} catch (_: URISyntaxException) {
					logger.error { "No valid CommunicationErrorAddress provided." }
					throw InvalidRedirectUrlException(ErrorTranslations.NO_URL)
				}
			} else {
				logger.error { "No CommunicationErrorAddress to perform a redirect provided." }
				throw InvalidRedirectUrlException(ErrorTranslations.NO_REDIRECT_AVAILABLE)
			}
		}
	}
}
